python/pythonUtilities.py

```diff
@@ -0,0 +1,177 @@
+import datetime
+import os
+import subprocess
+import sys
+
+
+# This contains the clicast command that was used to start the data server
+globalClicastCommand = ""
+
+# This global is used by all of the moudles to determine if they should
+# use the server logic or not.  This is set to True in the main() function
+# of vcastDataServer.py
+USE_SERVER = False
+
+# Key is the path to the environment, value is the process object
+# for the clicast instance for that environment
+clicastInstances = {}
+
+
+def setClicastInstance(enviroPath, processObject):
+    """
+    This function will set the clicast instance for the given environment
+    """
+    enviroPath = cleanEnviroPath(enviroPath)
+    clicastInstances[enviroPath] = processObject
+
+
+def removeClicastInstance(enviroPath):
+    """
+    This function will remove the clicast instance for the given environment
+    This should be called AFTER the process has been terminated
+    """
+    enviroPath = cleanEnviroPath(enviroPath)
+    if enviroPath in clicastInstances:
+        del clicastInstances[enviroPath]
+
+
+def startNewClicastInstance(enviroPath):
+    """
+    This function will start a new clicast instance and check
+    that it initializes correctly.  If it does we will return the
+    process object, if not we will return None
+    """
+
+    commandArgs = [globalClicastCommand, "-lc", "tools", "server"]
+    CWD = os.path.dirname(enviroPath)
+    processObject = subprocess.Popen(
+        commandArgs,
+        stdout=subprocess.PIPE,
+        stdin=subprocess.PIPE,
+        stderr=sys.stdout,
+        universal_newlines=True,
+        cwd=CWD,
+    )
+
+    # A valid clicast server start emits: "clicast-server-started"
+    # so we check for that, or we wait for the process to terminate
+    # which a non-server clicast version will do.
+    # Note that this is a failsafe because we should never get
+    # here if the clicast is not server capable.
+    while True:
+        responseLine = processObject.stdout.readline()
+        if responseLine.startswith("clicast-server-started"):
```

Comment:
Do we want to add a timeout for this check, or we're sure that the process will terminate in a reasonable time anyway?

Reply:
Good idea, the while true did make me a bit uncomfortable ...  improved with c4d4050